darwin/future/core/types/query.py

```diff
@@ -137,9 +184,17 @@ def __setitem__(self, index: int, value: T) -> None:
             self.results = list(self.collect())
         self.results[index] = value
 
-    @abstractmethod
-    def where(self, param: Param) -> Query[T]:
-        raise NotImplementedError("Not implemented")
+    def where(self, *args: object, **kwargs: str) -> Query[T]:
+        filters = QueryFilter._from_args(*args, **kwargs)
+        for item in filters:
+            self += item
+        return self
```

Comment:
So the problem was that for specific objects, like `DatasetQuery`, `where(...)` needs to call the constructor of `DatasetQuery` rather than `cls.init()` for mypy typing, all of our implementations basically just looked like 
```
class DatasetQuery(Query[DatasetMeta]):
    def where(self, ...) -> Query[T]
        filter = QueryFilter(...)
        return DatasetQuery(filter)
        
```

but using the `self.__class__()` in the `__add__()` (or anywhere else we want to implement this), that satisfies mypy's type restraint on the return, while removing a chunk of code we have to write for every query.